litr-filters/src/main/java/com/linkedin/android/litr/filter/audio/AudioOverlayFilter.kt

```diff
@@ -0,0 +1,211 @@
+package com.linkedin.android.litr.filter.audio
+
+import android.content.Context
+import android.media.MediaCodec
+import android.media.MediaFormat
+import android.net.Uri
+import com.linkedin.android.litr.codec.Decoder
+import com.linkedin.android.litr.codec.Frame
+import com.linkedin.android.litr.codec.MediaCodecDecoder
+import com.linkedin.android.litr.exception.TrackTranscoderException
+import com.linkedin.android.litr.filter.BufferFilter
+import com.linkedin.android.litr.io.MediaExtractorMediaSource
+import com.linkedin.android.litr.io.MediaSource
+import com.linkedin.android.litr.render.AudioProcessor
+import com.linkedin.android.litr.render.AudioProcessorFactory
+import com.linkedin.android.litr.transcoder.TrackTranscoder
+import com.linkedin.android.litr.utils.ByteBufferPool
+import java.nio.ByteBuffer
+import java.util.concurrent.LinkedBlockingDeque
+import kotlin.math.ceil
+import kotlin.math.min
+
+private const val BYTES_PER_SAMPLE = 2 // android uses 2 bytes per audio sample
+private const val UNDEFINED_VALUE = -1
+
+private const val TAG = "AudioOverlayFilter"
+
+class AudioOverlayFilter(
+    private val mediaSource: MediaSource,
+    private val decoder: Decoder
+) : BufferFilter {
+
+    constructor(context: Context, uri: Uri) : this(
+        MediaExtractorMediaSource(context, uri),
+        MediaCodecDecoder()
+    )
+
+    private val overlayChannelCount: Int
+    private val overlaySampleRate: Int
+
+    private val renderQueue = LinkedBlockingDeque<ByteBuffer>()
+    private val bufferPool = ByteBufferPool(true)
+    private val audioProcessorFactory = AudioProcessorFactory()
+    private var audioProcessor: AudioProcessor? = null
+
+    private var overlayTrack = 0
+    private var channelCount = UNDEFINED_VALUE
+    private var sampleRate = UNDEFINED_VALUE
+    private var samplingRatio = 1.0
+
+    init {
+        val mediaFormats = mutableListOf<MediaFormat>()
+        for (track in 0 until mediaSource.trackCount) {
+            mediaFormats.add(track, mediaSource.getTrackFormat(track))
+        }
+
+        overlayTrack = mediaFormats.indexOfFirst { mediaFormat ->
+            mediaFormat.containsKey(MediaFormat.KEY_MIME) &&
+                mediaFormat.getString(MediaFormat.KEY_MIME)?.startsWith("audio") == true
+        }
+
+        if (overlayTrack < 0) throw IllegalArgumentException("Audio overlay does not have an audio track")
```

Comment:
JC: how likely is it for the overlay source to have multiple audio tracks? I'm thinking of multiple languages or something like that, guessing this isn't common but just curious